src/pretix/control/templatetags/build_locale_url.py

```diff
@@ -0,0 +1,20 @@
+import urllib.parse
+
+from django import template
+
+register = template.Library()
+
+
+@register.filter
+def build_locale_url(request, locale_code):
+    """
+    Constructs the locale change URL by appending the safely encoded locale
+    and next path with the existing query string in a safe and automatic manner.
+    """
+    encoded_locale = urllib.parse.quote(locale_code)
+    params = {'locale': encoded_locale, 'next': request.path}
+    if request.META.get('QUERY_STRING'):
+        existing_params = urllib.parse.parse_qs(request.META['QUERY_STRING'])
```

Comment:
`request.GET` already contains the `dict` you want. But it looks like you want to override the `next` key if doesn't exist. If so, the function can be as simple as:


```py
params = request.GET.copy()  # Do copy because request.GET is immutable
params.setdefault('next', request.path)
params['locale'] = locale_code

return f'{request.path}?{params.urlencode()}'
```

Though I don't understand why you have to set default `next` with the same `request.path`? Will it cause circular redirection?